pkg/kubelet/nodeconfig/controller.go

```diff
@@ -0,0 +1,369 @@
+		// it probably indicates some sort of corruption
+		finalConfig = cc.badRollback(curUID, fmt.Sprintf(curFailLoadReasonFmt, curUID), fmt.Sprintf("error: %v", err))
+		return
+	}
+
+	// verify the integrity of the configuration we just loaded
+	toParse, err := toVerify.verify()
+	if err != nil {
+		finalConfig = cc.badRollback(curUID, fmt.Sprintf(curFailVerifyReasonFmt, curUID), fmt.Sprintf("error: %v", err))
+		return
+	}
+
+	// parse the configuration we just loaded into a KubeletConfiguration
+	cur, err := toParse.parse()
+	if err != nil {
+		finalConfig = cc.badRollback(curUID, fmt.Sprintf(curFailParseReasonFmt, curUID), fmt.Sprintf("error: %v", err))
+		return
+	}
+
+	// validate current config
+	if err := validateConfig(cur); err != nil {
+		finalConfig = cc.badRollback(curUID, fmt.Sprintf(curFailValidateReasonFmt, curUID), fmt.Sprintf("error: %v", err))
+		return
+	}
+
+	// check for crash loops if we're still in the trial period
+	if cc.curInTrial(cur.ConfigTrialDuration.Duration) {
+		if cc.crashLooping(*cur.CrashLoopThreshold) {
+			finalConfig = cc.badRollback(curUID, fmt.Sprintf(curFailCrashLoopReasonFmt, curUID), "")
+			return
+		}
+	} else if !cc.curIsLkg() {
+		// when the trial period is over, the current config becomes the last-known-good
+		cc.setCurAsLkg()
+	}
+
+	// update the status to note that we will use the current config
+	cc.setConfigOK(fmt.Sprintf(curRemoteMessageFmt, curUID), curRemoteOKReason, apiv1.ConditionTrue)
+	finalConfig = cur
+	return
+}
+
+// StartSyncLoop launches the sync loop that watches the API server for configuration updates
+// The `client` passed to StartSyncLoop will replace the controller's current client
+func (cc *NodeConfigController) StartSyncLoop(client clientset.Interface, nodeName string) {
+	if client != nil {
+		cc.client = client
+		cc.nodeName = nodeName
+		infof("starting sync loop")
+
+		// start the configuration sync loop
+		go func() {
+			defer utilruntime.HandleCrash()
+			fieldselector := fmt.Sprintf("metadata.name=%s", cc.nodeName)
+
+			// Add some randomness to resync period, which can help avoid controllers falling into lock-step
+			minResyncPeriod := 30 * time.Second
+			factor := rand.Float64() + 1
+			resyncPeriod := time.Duration(float64(minResyncPeriod.Nanoseconds()) * factor)
+
+			lw := &cache.ListWatch{
+				ListFunc: func(options metav1.ListOptions) (kuberuntime.Object, error) {
+					return cc.client.Core().Nodes().List(metav1.ListOptions{
+						FieldSelector: fieldselector,
+					})
+				},
+				WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+					return cc.client.Core().Nodes().Watch(metav1.ListOptions{
+						FieldSelector:   fieldselector,
+						ResourceVersion: options.ResourceVersion,
+					})
+				},
+			}
+
+			handler := cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					cc.onAddNodeEvent(obj)
+				},
+				UpdateFunc: func(old interface{}, nw interface{}) {
+					cc.onWatchNodeEvent(nw)
```

Comment:
Use a workqueue and append and dedup these. I don't like blocking so much in the informer loop. It seems like these could pile up since we are doing a lot of work in onWatchNodeEvent.